Sources/Basic/FileSystem.swift

```diff
@@ -467,5 +466,60 @@ public class InMemoryFileSystem: FileSystem {
     }
 }
 
+/// A rerooted view on an existing FileSystem.
+///
+/// This is a simple wrapper which creates a new FileSystem view into a subtree
+/// of an existing filesystem. This is useful for passing to clients which only
+/// need access to a subtree of the filesystem but should otherwise remain
+/// oblivious to its concrete location.
+public struct RerootedFileSystemView: FileSystem {
+    /// The underlying file system.
+    private var underlyingFileSystem: FileSystem
+
+    /// The root path within the containing file system.
+    private let root: AbsolutePath
+    
+    public init(_ underlyingFileSystem: inout FileSystem, rootedAt root: AbsolutePath) {
+        self.underlyingFileSystem = underlyingFileSystem
+        self.root = root
+    }
+
+    /// Adjust the input path for the underlying file system.
+    private func formUnderlyingPath(_ path: AbsolutePath) -> AbsolutePath {
+        if path == AbsolutePath.root {
+            return root
+        } else {
+            // FIXME: Optimize?
+            return root.appending(RelativePath(String(path.asString.characters.dropFirst(1))))
```

Comment:
I have actually long been of the opinion that a leading path separator should be allowed when the caller is specifically appending a subpath to an existing path, as is done in rerooting cases (e.g. a build system that appends `$(INSTALL_DIR)` to `$(INSTALL_ROOT)`).  I'm fine if the default behavior is more strict, but there should be some way, I think, to allow inputs to be a bit more flexible.